enclave-runtime/src/sync.rs

```diff
@@ -0,0 +1,104 @@
+//! Primitives to handle multithreaded state access in the enclave.
+//!
+//! Note: In general the design should try to minimize usage of these, as potential deadlocks can
+//! occur. Documentation of the `SgxRwLock` says that panics __might__ occur when trying to acquire
+//! a lock multiple times in the same thread. However, tests have shown that it also might result in
+//! a deadlock.
+//!
+//! @clangenb: Does currently not see any way to entirely get rid of these synchronization
+//! primitives because we can only start new threads from the untrusted side. `parking_lot` would be
+//! an alternative to consider for the primitives. It has several performance and ergonomic benefits
+//! over the `std` lib's. One of the benefits would be compile-time deadlock detection (experimental).
+//! Unfortunately, it would need to be ported to SGX.
+//!
+//! `https://amanieu.github.io/parking_lot/parking_lot/index.html`
+
+use crate::error::{Error, Result as EnclaveResult};
+use lazy_static::lazy_static;
+use std::sync::{SgxRwLock, SgxRwLockReadGuard, SgxRwLockWriteGuard};
+
+lazy_static! {
+	pub static ref LIGHT_CLIENT_LOCK: SgxRwLock<()> = Default::default();
+	pub static ref SIDECHAIN_DB_LOCK: SgxRwLock<()> = Default::default();
```

Comment:
So these locks are to guard against concurrent file access of the light client and side chain DB files?

Reply:
Yes, I am absolutely not sure if this suitable approach. I had a hard time finding material on the internet.

Reply:
It seems to get the job done, so let's keep it. If we discover issues with it or find a better solution, we can always come back and refactor it.